*[ft] Add Rollout Percentage functionality

chiraag918 opened the pull request.

floydkim commented:
Hello! @chiraag918 
Thank you for implementing such a great feature.

Although I haven’t received any direct requests for it yet, I believe there are definitely use cases where the progressive rollout feature will be useful.
It’s been a few days since you submitted the PR, but I’ve been reviewing the code whenever I get the chance.

--

#### About the feature

The idea of using the device ID for bucketing is very interesting!
However, as far as I understand, the rollout percentage directly corresponds to the bucket number.
This means that devices in the smaller-numbered buckets will always be part of the early rollout, while those in larger-numbered buckets will always receive updates later.
In other words, certain user groups could repeatedly experience unstable releases.

I’ve been thinking about how we could add a bit of randomness to this.
One idea is to generate the hash not only from the deviceId, but also by including the package’s label or packageHash values.
What do you think? Would you be open to accepting this idea?

--
#### About the `disabled` flag

I’m not entirely convinced about the need for the `disabled` flag.
In what use cases would it be necessary?

Also, I couldn’t find the corresponding implementation in the diff. (unless I missed it)
I’d prefer that it not be included in the diff until I better understand its usefulness. 🙏 
If it turns out to be needed, it would be better to implement it in a separate PR.

floydkim commented:
@chiraag918 
Thank you for accepting my idea 😄 

I noticed that you added a random factor along with the package hash.
However, this means that when a user deletes and reinstalls the app, they could end up in a different bucket, and there’s a chance they won’t be able to return to the app version they had before deleting it.
I believe this has both pros and cons. Personally, I think it appealing that our library could be distinguished by ensuring users maintain the same experience with the same update.

I also think that adding a random factor somewhat diminishes the meaning of using the device ID and package hash. With the random factor added, doesn’t the outcome essentially become the same as just using Math.random()? 🤔 

If the random factor were removed, I believe deterministic bucketing could be achieved for the same device and the same CodePush update.

(If my previous comment using the phrase “a bit of randomness” caused any confusion, I sincerely apologize. 🙏)

chiraag918 commented:
@floydkim true, if the user were to uninstall & reinstall the app, the previous decision can get changed depending on the random factor 😅. So in that case, keeping device ID and packageHash would solve for this use case right? I added the random factor to make it a bit more random. I can remove the random multiplier. Any other suggestions or thoughts? 

(No worries, these conversations helps me with understanding these corner cases, thanks 😀)

floydkim commented:
@chiraag918 
I think it would be better to remove the random factor and only use the device ID and package hash.

If the need ever arises, we could later add an option to make the bucketing completely random. :)

chiraag918 commented:
@floydkim I've removed the random factor multiplier in my recent commit. Request you to please review the PR once. Thank you.

floydkim commented on CodePush.js:
```diff
@@ -493,6 +559,11 @@ async function syncInternal(options = {}, syncStatusChangeCallback, downloadProg
       return CodePush.SyncStatus.UPDATE_INSTALLED;
     };
 
+    if(remotePackage?.skipRollout){
+      syncStatusChangeCallback(CodePush.SyncStatus.UNKNOWN_ERROR);
+      return CodePush.SyncStatus.UNKNOWN_ERROR;
+    }
```
Is there a particular reason why we need to distinguish the state where it’s skipped because it’s not within the rollout range?

In apps that used the UP_TO_DATE state to inform users that "the app is up to date", code changes will be necessary due to the addition of this new state.
(If not using the rollout feature, no changes would be needed, of course.)

There may be operational or management reasons, but I think it’s possible to collect that information through the onRolloutSkipped callback.

chiraag918 replied:
Created a different state to maintain a different state for codepush skipped scenario for better readability during dev. Instead what would you suggest to keep this state as?

chiraagnm replied:
Okay makes sense, have updated this state to UP_TO_DATE and removed CODEPUSH_SKIPPED refs.

chiraag918 replied:
Okay makes sense, have changed this state to UP_TO_DATE and have removed refs for CODEPUSH_SKIPPED.

floydkim commented on android/app/src/main/java/com/microsoft/codepush/react/RolloutStorageModule.java:
```diff
@@ -0,0 +1,38 @@
+import com.facebook.react.bridge.ReactMethod;
+import com.facebook.react.bridge.Promise;
+
+public class RolloutStorageModule extends ReactContextBaseJavaModule {
```
The SettingsManager already handles interactions with SharedPreference.
Was the reason you added a new module instead to separate concerns?

floydkim replied:
Oh, no. I was just curious about the reason.
For now, since I’m not familiar with the implementation of SettingsManager, I like the way you implemented.

chiraag918 replied:
Okay great!

floydkim commented on cli/commands/releaseCommand/index.js:
```diff
@@ -16,6 +16,7 @@ program.command('release')
     .option('-j, --js-bundle-name <string>', 'JS bundle file name (default-ios: "main.jsbundle" / default-android: "index.android.bundle")')
     .option('-m, --mandatory <bool>', 'make the release to be mandatory', parseBoolean, false)
     .option('--enable <bool>', 'make the release to be enabled', parseBoolean, true)
+    .option('--rollout <number>', 'rollout percentage (0-100)', parseFloat, 100)
```
Personally, I think it would be better not to add the rollout property when the rollout feature is not used, so that the JSON data doesn’t grow unnecessarily.

chiraagnm replied:
Oops my bad. Yes reverted this change and now there's no default argument in option('--rollout....') call.

floydkim commented on ios/CodePush/RolloutStorage.h:
```diff
@@ -0,0 +1,9 @@
+
+@interface RCT_EXTERN_MODULE(RolloutStorage, NSObject)
+
+RCT_EXTERN_METHOD(setItem:(NSString *)key value:(NSString *)value)
+RCT_EXTERN_METHOD(getItem:(NSString *)key resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
+RCT_EXTERN_METHOD(removeItem:(NSString *)key)
```
I’m not sure if the RCT_EXPORT_METHOD macro is needed in the header file as well.

From what I can find, it doesn’t seem to be commonly declared in headers.
- https://github.com/search?type=code&q=RCT_EXPORT_METHOD+language%3AObjective-C+path%3A*.h
- https://github.com/search?type=code&q=RCT_EXPORT_METHOD+language%3AObjective-C+path%3A*.m
- https://github.com/search?type=code&q=RCT_EXPORT_METHOD+language%3AObjective-C%2B%2B+path%3A*.mm

floydkim replied:
I’m not an iOS/Objective-C expert, so I'd like to follow either the library’s current implementation style or approaches used in other libraries.

In the CodePush.h and CodePush.m files, the RCT_EXPORT_METHOD macro appears only in the .m file, and the header doesn’t declare these methods.
(I’m not certain if this is because the macro removes the need for function prototype declarations.. 🤔)

chiraag918 replied:
From my understanding, RCT_EXPORT_METHOD need not be included in the header files. The ones that you shared seems to be like react bridging header files to bridge objective-c & swift code, since, swift doesn't yet support exposing native methods to React JS like how we do in objective-c. 

1. Similarly RCT_EXTERN_METHOD is similar to RCT_EXPORT_METHOD but used with swift files. Including/Excluding it in our case doesn't make a difference is what I suppose. 
2. Yes you're right, in CodePush.h and CodePush.m files, the RCT_EXPORT_METHOD macro appears only in the .m file, since its not required to duplicate the macro in both the files. 

But considering both 1 & 2, I've removed RCT_EXTERN_METHOD from the .h file and RCT_EXPORT_METHOD only is used in the .m file, as this is what RN suggests: https://reactnative.dev/docs/0.73/native-modules-ios#create-custom-native-module-files

floydkim replied:
Oh, I confused [RCT_EXTERN_METHOD](https://github.com/facebook/react-native/blob/35bee1a8572ddc3e6480e4e8c06bc40ede15929e/packages/react-native/React/Base/RCTBridgeModule.h#L253-L281) with [RCT_EXPORT_METHOD](https://github.com/facebook/react-native/blob/35bee1a8572ddc3e6480e4e8c06bc40ede15929e/packages/react-native/React/Base/RCTBridgeModule.h#L166-L206).

I wasn’t familiar with RCT_**EXTERN**_MODULE, but as you mentioned, it’s used to register implementations in Swift.
Thanks for sharing your knowledge.

Anyway, the final code you’ve updated looks good 👍

floydkim reviewed (commented):
Hi @chiraag918 !
Sorry for the delay.

I reviewed the code thoroughly and had some questions, so I’ve asked them.
Please respond when you have time. 😄 

Once I understand your answers, I’m thinking of merging the PR first and then making the changes myself, including some minor coding style tweaks.
What do you think about it?
I want to reduce the back-and-forth on feedback and move quickly toward releasing the feature.

chiraag918 commented on typings/react-native-code-push.d.ts:
```diff
@@ -342,9 +347,11 @@ declare namespace CodePush {
      * @param handleBinaryVersionMismatchCallback An optional callback for handling target binary version mismatch
      */
-    function checkForUpdate(handleBinaryVersionMismatchCallback?: HandleBinaryVersionMismatchCallback): Promise<RemotePackage | null>;
+    function checkForUpdate(deploymentKey?: string, handleBinaryVersionMismatchCallback?: HandleBinaryVersionMismatchCallback): Promise<RemotePackage | null>;
```
Sorry took this change from the package code that I was using to test my changes. Have reverted this change.

chiraag918 commented on CodePush.js:
```diff
@@ -493,6 +563,11 @@ async function syncInternal(options = {}, syncStatusChangeCallback, downloadProg
+    if(remotePackage?.skipRollout){
+      syncStatusChangeCallback(CodePush.SyncStatus.UNKNOWN_ERROR);
+      return CodePush.SyncStatus.UNKNOWN_ERROR;
+    }
```
Updated this

chiraag918 commented on CodePush.js:
```diff
@@ -163,7 +222,14 @@ async function checkForUpdate(handleBinaryVersionMismatchCallback = null) {
     return null;
   } else {
-    const remotePackage = { ...update, ...PackageMixins.remote() };
+    const remotePackage = { ...PackageMixins.remote(), ...update };
```
Hmmm I don't think so, initially I guess it was taking default rollback value as 100, hence, had changed the order, so the rollout value comes from the remote fetched config. Now I've removed the fallback value as suggested in one of the other other PR comments.

chiraag918 commented on CodePush.js:
```diff
@@ -6,8 +6,66 @@ import { SemverVersioning } from './versioning/SemverVersioning'
+  if(prevRolloutCacheKey)
+    await RolloutStorage.removeItem(prevRolloutCacheKey);
+
+  await RolloutStorage.setItem(ROLLOUT_CACHE_KEY, rolloutKey);
```
Okay pls do share once you refactor.

chiraag918 commented on CodePush.js:
```diff
@@ -6,8 +6,66 @@ import { SemverVersioning } from './versioning/SemverVersioning'
+}
+
+function getRolloutKey(label, rollout) {
+  return `${ROLLOUT_CACHE_PREFIX}${label}_rollout_${rollout ?? 'full'}`;
```
Have made this change.

chiraag918 commented:
Sorry was little busy with work, hence the delay. I've addressed the PR comments. I think 3 comments, I wasn't sure, need your suggestion on it. Pls go through them and lmk, will take action. Thank you!!!

floydkim commented:
@chiraag918 Thank you for your great work.
I think we can merge this PR now, and then I will do some testing and minor code style refactoring.
(I'll open a PR and mention you so you can read through the changes.)

I'll slightly update the PR title for the GitHub release notes and then merge it.
Thank you for bearing with the communication overhead and for your understanding.

chiraag918 commented:
@floydkim Thank you so much for your time, effort, for guiding this through and accepting my contribution. I really appreciate the collaboration. Please feel free to reach out if anything comes up during your testing or refactoring, I’ll be happy to help with any issues 😁
